libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -377,20 +419,29 @@ void AC_PosControl::set_alt_target_from_climb_rate_ff(float climb_rate_cms, floa
     // jerk_z is calculated to reach full acceleration in 1000ms.
     float jerk_z = accel_z_cms * POSCONTROL_JERK_RATIO;
 
-    float accel_z_max = MIN(accel_z_cms, safe_sqrt(2.0f * fabsf(_vel_desired.z - climb_rate_cms) * jerk_z));
+    float accel_z_max = MIN(accel_z_cms, safe_sqrt(2.0f * fabsf(climb_rate_cms- _vel_desired.z) * jerk_z));
 
+    // jerk limit the acceleration increase
     _accel_last_z_cms += jerk_z * dt;
-    _accel_last_z_cms = MIN(accel_z_max, _accel_last_z_cms);
+    // jerk limit the decrease as zero error is approached
+    _accel_last_z_cms = MIN(_accel_last_z_cms, accel_z_max);
+    // remove overshoot during last time step
+    _accel_last_z_cms = MIN(_accel_last_z_cms, fabsf(climb_rate_cms- _vel_desired.z) / dt);
+
+    if (is_positive(climb_rate_cms- _vel_desired.z)){
+        _accel_desired.z = _accel_last_z_cms;
+    } else {
+        _accel_desired.z = -_accel_last_z_cms;
+    }
 
     float vel_change_limit = _accel_last_z_cms * dt;
     _vel_desired.z = constrain_float(climb_rate_cms, _vel_desired.z - vel_change_limit, _vel_desired.z + vel_change_limit);
-    _flags.use_desvel_ff_z = true;
 
     // adjust desired alt if motors have not hit their limits
     // To-Do: add check of _limit.pos_down?
     if ((_vel_desired.z < 0 && (!_motors.limit.throttle_lower || force_descend)) || (_vel_desired.z > 0 && !_motors.limit.throttle_upper && !_limit.pos_up)) {
-        _pos_target.z += _vel_desired.z * dt;
     }
+    _pos_target.z += _vel_desired.z * dt;
```

Comment:
The above if (line 442) seems to not be doing anything anymore?

Reply:
fixed